provider/webhook/httpapi.go

```diff
@@ -0,0 +1,143 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package webhook
+
+import (
+	"context"
+	"encoding/json"
+	"net"
+	"net/http"
+	"time"
+
+	"sigs.k8s.io/external-dns/endpoint"
+	"sigs.k8s.io/external-dns/plan"
+	"sigs.k8s.io/external-dns/provider"
+
+	log "github.com/sirupsen/logrus"
+)
+
+type WebhookServer struct {
+	provider provider.Provider
+}
+
+func (p *WebhookServer) recordsHandler(w http.ResponseWriter, req *http.Request) {
+	switch req.Method {
+	case http.MethodGet:
+		records, err := p.provider.Records(context.Background())
+		if err != nil {
+			log.Errorf("Failed to get Records: %v", err)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		w.Header().Set(contentTypeHeader, mediaTypeFormatAndVersion)
+		w.WriteHeader(http.StatusOK)
+		if err := json.NewEncoder(w).Encode(records); err != nil {
+			log.Errorf("Failed to encode records: %v", err)
+		}
+		return
+	case http.MethodPost:
+		var changes plan.Changes
+		if err := json.NewDecoder(req.Body).Decode(&changes); err != nil {
+			log.Errorf("Failed to decode changes: %v", err)
+			w.WriteHeader(http.StatusBadRequest)
+			return
+		}
+		err := p.provider.ApplyChanges(context.Background(), &changes)
+		if err != nil {
+			log.Errorf("Failed to Apply Changes: %v", err)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		w.WriteHeader(http.StatusNoContent)
+		return
+	default:
+		log.Errorf("Unsupported method %s", req.Method)
+		w.WriteHeader(http.StatusBadRequest)
+	}
+}
+
+func (p *WebhookServer) adjustEndpointsHandler(w http.ResponseWriter, req *http.Request) {
+	if req.Method != http.MethodPost {
+		log.Errorf("Unsupported method %s", req.Method)
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	pve := []*endpoint.Endpoint{}
+	if err := json.NewDecoder(req.Body).Decode(&pve); err != nil {
+		log.Errorf("Failed to decode in adjustEndpointsHandler: %v", err)
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+	w.Header().Set(contentTypeHeader, mediaTypeFormatAndVersion)
+	pve = p.provider.AdjustEndpoints(pve)
+	if err := json.NewEncoder(w).Encode(&pve); err != nil {
+		log.Errorf("Failed to encode in adjustEndpointsHandler: %v", err)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+}
+
+func (p *WebhookServer) negotiateHandler(w http.ResponseWriter, req *http.Request) {
+	b, err := p.provider.GetDomainFilter().MarshalJSON()
+	if err != nil {
+		log.Errorf("Failed to marshal domain filter: %v", err)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	w.Header().Set(contentTypeHeader, mediaTypeFormatAndVersion)
+	w.Write(b)
```

Comment:
Instead of explicitly calling `MarshalJSON()` above, you could simply `json.NewEncoder(w).Encode(&domainFilter)`. That would remove knowledge that the JSON serialization is implemented through a `MarshalJSON()` receiver.